### PR TITLE
[WFLY-18237]: Adding a connector shouldn't require to reload.

### PR DESCRIPTION
**Pull request by ehsavoie**
* Adding a connector dynamically to the broker to avoid the reload requirement.

Jira: https://issues.redhat.com/browse/WFLY-18237

### CODE REVIEW

File: messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ConnectorAdd.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    protected boolean requiresRuntime(OperationContext context) {
+        return context.isDefaultRequiresRuntime() && !context.isBooting();
+    }
+
+    @Override
+    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+        if (MessagingServices.isSubsystemResource(context)) {
+            return;
+        }
+        if (isServiceInstalled(context)) {
```
Comment (bstansberry) [inline messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ConnectorAdd.java:48]: What about the case where we are not booting and the service isn't installed?

File: messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/TransportConfigOperationHandlers.java [hunk 2]
```diff
@@ -340,16 +346,96 @@ static void processAcceptors(final OperationContext context, final Configuration
...
+        CapabilityServiceBuilder builder = context.getCapabilityServiceTarget().addCapability(CONNECTOR_CAPABILITY);
+        builder.provides(CONNECTOR_CAPABILITY.getCapabilityServiceName(context.getCurrentAddress()));
+        builder.setInitialMode(ServiceController.Mode.ACTIVE);
+        switch (connectorType) {
+            case CommonAttributes.CONNECTOR: {
+                final Map<String, Object> parameters = getParameters(context, config, CONNECTORS_KEYS_MAP);
+                String socketBindingName = GenericTransportDefinition.SOCKET_BINDING.resolveModelAttribute(context, config).asStringOrNull();
+                final String clazz = FACTORY_CLASS.resolveModelAttribute(context, config).asString();
+                if (config.hasDefined(CommonAttributes.SSL_CONTEXT.getName())) {
+                    String sslContextName = config.get(CommonAttributes.SSL_CONTEXT.getName()).asString();
```
Comment (bstansberry) [inline messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/TransportConfigOperationHandlers.java:360]: The 'config' var is actual the 'operation' node from ConnectorAdd. We shouldn't directly use the operation node in Stage.RUNTIME.

It should use resolveModelAttribute as well.

File: messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/TransportConfigOperationHandlers.java [hunk 3]
```diff
@@ -340,16 +346,96 @@ static void processAcceptors(final OperationContext context, final Configuration
...
+            }
+            case CommonAttributes.REMOTE_CONNECTOR: {
+                final Map<String, Object> parameters = getParameters(context, config, CONNECTORS_KEYS_MAP);
+                String socketBindingName = GenericTransportDefinition.SOCKET_BINDING.resolveModelAttribute(context, config).asStringOrNull();
+                if (!config.hasDefined(FACTORY_CLASS.getName())) {
+                    config.get(FACTORY_CLASS.getName()).set(NettyConnectorFactory.class.getName());
+                }
+                final String clazz = FACTORY_CLASS.resolveModelAttribute(context, config).asString();
+                if (config.hasDefined(CommonAttributes.SSL_CONTEXT.getName())) {
+                    String sslContextName = config.get(CommonAttributes.SSL_CONTEXT.getName()).asString();
```
Comment (bstansberry) [inline messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/TransportConfigOperationHandlers.java:381]: Use the model, not the operation, use resolveModelAttribute().

File: messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/TransportConfigOperationHandlers.java [hunk 4]
```diff
@@ -340,16 +346,96 @@ static void processAcceptors(final OperationContext context, final Configuration
...
+                // ARTEMIS-803 Artemis knows that is must not offset the HTTP port when it is used by colocated backups
+                parameters.put(TransportConstants.HTTP_UPGRADE_ENABLED_PROP_NAME, true);
+                parameters.put(TransportConstants.HTTP_UPGRADE_ENDPOINT_PROP_NAME, HTTPConnectorDefinition.ENDPOINT.resolveModelAttribute(context, config).asString());
+                ModelNode serverNameModelNode = HTTPConnectorDefinition.SERVER_NAME.resolveModelAttribute(context, config);
+                // use the name of this server if the server-name attribute is undefined
+                String serverName = serverNameModelNode.isDefined() ? serverNameModelNode.asString() : MessagingServices.getActiveMQServerPathAddress(context.getCurrentAddress()).getLastElement().getValue();
+                parameters.put(ACTIVEMQ_SERVER_NAME, serverName);
+                if (config.hasDefined(CommonAttributes.SSL_CONTEXT.getName())) {
+                    String sslContextName = config.get(CommonAttributes.SSL_CONTEXT.getName()).asString();
```
Comment (bstansberry) [inline messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/TransportConfigOperationHandlers.java:400]: Use the model, not the operation; use resolveModelAttribute()

**Review by bstansberry — CHANGES_REQUESTED** [review]

File: messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ConnectorAdd.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            } catch (CapabilityServiceSupport.NoSuchCapabilityException ex) {
+                throw new OperationFailedException(ex);
+            }
+        }
+    }
+
+    @Override
+    protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource) throws OperationFailedException {
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
+    }
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ConnectorAdd.java:60]: no, leftover from debugging

File: messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ConnectorAdd.java (hunk 1 above)
Comment (ehsavoie) [reply messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ConnectorAdd.java:50]: maybe I can move it to requiresRuntime too

**Review by bstansberry — APPROVED** [review]

**Comment by bstansberry** [comment]
Thanks @ehsavoie

**Pull request merged**